webapp/src/cmds.ts

```diff
@@ -450,7 +464,128 @@ export function showDisconnectAsync(): Promise<void> {
+    if (!supported) return;
+
+    if (navigator?.serviceWorker?.controller) {
+        return new Promise<void>((resolve, reject) => {
+            pendingPacketIOLockResolver = resolve;
+            pendingPacketIOLockRejecter = reject;
+            pxt.log("[CLIENT]: requesting lock " + lockRef)
+            sendServiceWorkerMessage({
+                type: "serviceworkerclient",
+                action: "request-packet-io-lock",
+                lock: lockRef
+            });
+        })
+        .finally(() => {
+            pendingPacketIOLockResolver = undefined;
+            pendingPacketIOLockRejecter = undefined;
+        })
+    }
+}
+
+function sendServiceWorkerMessage(message: pxt.ServiceWorkerClientMessage) {
+    if (navigator?.serviceWorker?.controller) {
+        navigator.serviceWorker.controller.postMessage(message);
+    }
+}
+
+export async function handleServiceWorkerMessageAsync(message: pxt.ServiceWorkerMessage) {
+    if (message.action === "packet-io-lock-disconnect" && !pendingPacketIOLockResolver && lockRef === message.lock) {
+        pxt.log("[CLIENT]: received disconnect request " + lockRef)
+
+        if (deployingPacketIO || pxt.packetio.isConnecting()) {
+            sendServiceWorkerMessage({
+                type: "serviceworkerclient",
+                action: "packet-io-lock-disconnect",
+                lock: lockRef,
+                didDisconnect: false
+            });
+        }
+        else {
+            await disconnectAsync();
+        }
+    }
+    else if (message.action === "packet-io-lock-granted" && message.lock === lockRef && pendingPacketIOLockResolver) {
+        if (message.granted) {
+            pxt.log("[CLIENT]: received granted lock " + lockRef)
+            pendingPacketIOLockResolver();
+            hasLock = true;
+        }
+        else {
+            pxt.log("[CLIENT]: received denied lock " + lockRef)
+            pendingPacketIOLockRejecter();
+        }
+        pendingPacketIOLockResolver = undefined;
+        pendingPacketIOLockRejecter = undefined;
+    }
+    else if (message.action === "packet-io-supported" && serviceWorkerSupportedResolver) {
+        serviceWorkerSupportedResolver();
+        serviceWorkerSupported = true;
+        serviceWorkerSupportedResolver = undefined;
+    }
+    else if (message.action === "packet-io-status") {
+        sendServiceWorkerMessage({
+            type: "serviceworkerclient",
+            action: "packet-io-status",
+            hasLock: hasLock,
+            lock: lockRef
+        });
+    }
+}
+
+async function checkIfServiceWorkerSupportedAsync() {
+    if (serviceWorkerSupported !== undefined) return serviceWorkerSupported;
+
+    const p =  new Promise<void>(resolve => {
+        serviceWorkerSupportedResolver = resolve;
+        sendServiceWorkerMessage({
+            type: "serviceworkerclient",
+            action: "packet-io-supported"
+        });
+    });
+
+    try {
+        await pxt.U.promiseTimeout(1000, p);
```

Comment:
Making sure I follow here as it took me a sec -- we're effectively making a deferred above with a timeout here, where `serviceWorkerSupportedResolver` is being kept for `handleServiceWorkerMessage` to resolve w/in 1 second?

Reply:
yep! i'll add a comment to clarify. this is just to handle the case where the new service worker has been downloaded, but hasn't activated yet